### PR TITLE
Rounding fixes/line item rounding

### LINKED ISSUE #17970: Rounding issue in WC 3.2.5
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

1. Store is set to enter product prices excluding tax
2. Front end displays prices including tax (typical for UK retail store)
3. Tax is 20% (UK vat) - applied to all products
4. Add 2 products whose retail prices (inc tax) are: £65, £50. Their net prices are therefore: £54.1666667 and £41.6666667. At this point, the products display the correct, rounded price of £65 and £50
5. Add both products to the cart (free shipping)

## Expected/actual behavior

When I follow those steps, the grand total in the cart is £115.01.
It should of course be £115 exactly (£65 + £50).

**If nothing was rounded until the final total, we would have the correct final total:**
+ £54.1666667 (net of £65)
+ £41.6666667 (net of £50)
+ £10.8333333 (tax element of £65)
+ £8.3333333 (tax element of £50)
–––––––––––––––––––––––––––––––––––
TOTAL £115 (exactly with no rounding needed)

**Equally, if everything was rounded (net AND tax) per line, we also get to the correct total:**
+ £54.17 (rounded net of £65)
+ £41.67 (rounded net of £50)
+ £10.83 (rounded tax of £65)
+ £8.33 (rounded tax of £50)
–––––––––––––––––––––––––––––––––––
TOTAL £115

**However it seems that each product net price is being rounded per line but the tax is not. The tax total is then rounded before being summed with the net product total - That's the only way I can get to a final total of £115.01 so I guess is where the issue is.**

_PRODUCTS_
+ £54.17 (rounded net of £65)
+ £41.67 (rounded net of £50)
–––––––––––––––––––––––––––––––––––
Net product total **£95.84**

_TAX_
+ £10.8333333 (non-rounded tax of £65)
+ £8.3333333 (non-rounded tax element of £50)
–––––––––––––––––––––––––––––––––––
Tax total **£19.1666666666**

Grand total if tax total is rounded before summing tax with net total: 
£115.01 (£95.84 + £19.17)

Grand total if tax total is left unrounded before summing with net total:
£115 (£95.84 + £19.6666667, then rounded) 

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin here.
```
</details>

### PR DESCRIPTION
**Pull request by mikejolley**
This PR attempts to address some bugs reported in #17970 relating to rounding (in this case, excl tax prices using high precision, more than store DP setting).

Closes #17970 

This also addresses some of the concerns in #17863 - rounding per line was broken/not doing anything before these changes. Now, the setting is once again doing something, even with cent values.

Some notes;

- We store unrounded totals/subtotals per line item. This is so when you ‘recalculate’ via admin, the taxes will remain the same.
- Unless rounding at subtotal level, rounding will happen before totalling up taxes.
- Rounding will also occur when summing line item totals.
- Adding precision will use cents but with added precision as configured.

cc @claudiulodro @ragulka

Needs good testing before considering which version to merge to. I’ve assumed 3.3 for now.

### CODE REVIEW

**Comment by claudiulodro** [comment]
This looks good to me (and I assume tests will pass again after `round total to cents` commit). I'll write a few more tests for it tomorrow to make sure there aren't edge cases.

**Comment by claudiulodro** [comment]
I wasn't able to determine the PHP 5.2 issue. It seems like it has to be either something with `round` or `wc_round_tax_total` since it's isolated to PHP 5.2.

File: includes/class-wc-cart-totals.php [hunk 1]
```diff
@@ -827,4 +817,18 @@ protected function calculate_totals() {
 		// Allow plugins to filter the grand total, and sum the cart totals in case of modifications.
 		$this->cart->set_total( max( 0, apply_filters( 'woocommerce_calculated_total', $this->get_total( 'total' ), $this->cart ) ) );
 	}
+
+	/**
+	 * Apply rounding to an array of taxes before summing. Rounds to store DP setting, ignoing precision.
```
Comment (ragulka) [inline includes/class-wc-cart-totals.php]: typo - should be ignoring

**Comment by ragulka** [comment]
@mikejolley based on my testing, this fixes the issues described in #17863 - tested with both rounding at line level as well as subtotal level. I also opened a PR (#18054) with a passing test which failed prior to this change. Feel free to include that if you think that's helpful.

Based on my cart setup, these 3 modes work flawlessly, with no difference between cart contents total and cart total:
* prices include tax, rounding at line level
* prices include tax, rounding at subtotal level
* prices exclude tax, rounding at line level

When prices exclude tax and rounding at subtotal level (and displaying prices including tax in cart), there is a 0.01$ difference between the cart contents total (manually adding up cart items) and cart total. However, because adding up unrounded taxes, this is expected.

I guess shops that exclude tax in prices and round at subtotal level also tend to not display prices with taxes included in shop or cart, so this difference becomes invisible to the customer.

In conclusion - looks good.

**Comment by claudiulodro** [comment]
Looks good. I'll merge after tests are finished running.

**Pull request merged**